### PR TITLE
Efficient loftr

### PR DESCRIPTION
**Pull request by AbhiKhoyani**
## 📝 Description

**Fixes/Relates to:** [#3282](https://github.com/kornia/kornia/issues/3282)

---

## 🛠️ Changes Made
Add EfficientLoFTR implementation and related utilities
- Introduced  class for fine-level matching in ELoFTR.
- Added  for sinusoidal position encoding in 2D images.
- Implemented supervision utilities for coarse and fine-level supervision in ELoFTR.
- Refactored  in **LoFTR** to inherit from  for shared functionality. Reducing duplicate code.
- Created tests for EfficientLoFTR to validate pretrained models and output shapes.
---

## 🧪 How Was This Tested?
- [ ] **Unit Tests:** [test_eloftr.py](https://github.com/AbhiKhoyani/kornia/commit/202a902a2c4192e787c93cc1f280f094be7cc21f#diff-202352806618678239fba23967f0d148eb72fb81486f2c34fbe3825e94fd341a)

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [x] 🟢 **No AI used.**
- [ ] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [x] I am assigned to the linked issue (required before PR submission)
- [x] The linked issue has been approved by a maintainer
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

### CODE REVIEW

**Comment by ducha-aiki** [comment]
>        elif self.match_type == "sinkhorn":
            try:
                from .superglue import log_optimal_transport
            except ImportError:
                raise ImportError("download superglue.py first!") from None
                
 Please check the code carefully and remove such things. kornia should have the optimal transport in lightglue or loftr

**Comment by AbhiKhoyani** [comment]
> > ```
> >    elif self.match_type == "sinkhorn":
> > ```
> 
> ```
>         try:
>             from .superglue import log_optimal_transport
>         except ImportError:
>             raise ImportError("download superglue.py first!") from None
> ```
> 
> Please check the code carefully and remove such things. kornia should have the optimal transport in lightglue or loftr

On main branch itself its like tis only. I just refactored the code to reuse same part from loftr in eloftr. Please refer this [line](https://github.com/kornia/kornia/blame/main/kornia/feature/loftr/utils/coarse_matching.py#L104). However, I can make this changes after referring current kornia cobebase if there are any relevant functionality.

**Review by edgarriba — COMMENTED** [review]
Thanks for the persistence on this — the structure mirrors the existing LoFTR package nicely, the shared `BaseCoarseMatching` refactor with inlined Sinkhorn resolves the earlier superglue concern, and hosting the checkpoint on the kornia HF org solves the Google Drive problem from #3282. CI is green. A few things before merge:

1. Module filename typo: `effiicient_loftr.py` → `efficient_loftr.py` (it's an importable path).
2. `test_pretrained_outdoor` only compares shapes — the reference file already contains keypoints, so please compare values like `test_loftr.py` does.
3. `reparameter()` is defined but never invoked — upstream fuses the RepVGG branches before inference, which is where much of ELoFTR's speed comes from. Please either apply it after loading pretrained weights (verifying against the reference output) or document that the checkpoint is already fused.
4. `utils/supervision.py` is unused and uses upstream config keys incompatible with `default_cfg` — suggest dropping it.
5. Please add `EfficientLoFTR` to `docs/source/feature.rst` and an `ELoFTR2024` entry to `references.bib` — the docstring cites it, so the docs build will break otherwise.

We'll reopen #3282; please use `Fixes #3282` in the description so it links. Once this lands, if you're up for more model work, #3488 (Depth Anything 3) is high-demand and would benefit from exactly this kind of careful porting.

**Comment by AbhiKhoyani** [comment]
@edgarriba Updated this PR with your suggested comments. Fixes #3282

File: kornia/feature/loftr/utils/coarse_matching.py [hunk 1]
```diff
@@ -80,106 +80,33 @@ def compute_max_candidates(p_m0: torch.Tensor, p_m1: torch.Tensor) -> torch.Tens
     return max_cand
 
 
-class CoarseMatching(nn.Module):
-    """Perform coarse-level matching between two sets of feature maps.
-
-    This module calculates the confidence matrix and produces coarse matches
-    using an optimal transport layer or a dual-softmax operator.
+def log_sinkhorn_iterations(Z: torch.Tensor, log_mu: torch.Tensor, log_nu: torch.Tensor, iters: int) -> torch.Tensor:
+    """Perform Sinkhorn Normalization in Log-space for stability.
```
Comment (ducha-aiki) [inline kornia/feature/loftr/utils/coarse_matching.py:84]: Add sinkhorn doesn't mean "copy from SuperGlue w/o licence header" -- even the docstring is the same. 
https://github.com/magicleap/SuperGluePretrainedNetwork/blob/master/models/superglue.py#L143

Instead, as I asked for, please use OpenGlue implementation (https://github.com/ucuapps/OpenGlue/blob/main/models/superglue/optimal_transport.py ) with proper attribution from where it was taken, in comment.

File: kornia/feature/efficient_loftr/efficient_loftr.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+                "hw1_c": feat_c1.shape[2:],
+                "hw0_f": [feat_c0.shape[2] * mul, feat_c0.shape[3] * mul],
+                "hw1_f": [feat_c1.shape[2] * mul, feat_c1.shape[3] * mul],
+            }
+        )
+
+        # 2. coarse-level loftr module
+        mask_c0 = mask_c1 = None  # mask is useful in training
+        if "mask0" in data:
+            mask_c0, mask_c1 = data["mask0"], data["mask1"]
```
Comment (ducha-aiki) [inline kornia/feature/efficient_loftr/efficient_loftr.py:184]: What about resize? Kornia's LoFTR resizes resize(..., hw0_c, interpolation="nearest")

otherwise mask breaks the thing

File: kornia/feature/efficient_loftr/efficient_loftr.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+    Example:
+        >>> img1 = torch.rand(1, 1, 128, 128)
+        >>> img2 = torch.rand(1, 1, 128, 128)
+        >>> input = {"image0": img1, "image1": img2}
+        >>> eloftr = EfficientLoFTR()
+        >>> out = eloftr(input)
+    """
+
+    def __init__(self, config: dict[str, Any] = default_cfg, pretrained: Optional[str] = "outdoor") -> None:
```
Comment (ducha-aiki) [inline kornia/feature/efficient_loftr/efficient_loftr.py:103]: kornia loftr uses `LoFTR(pretrained="outdoor", config=...)

Let's switch arg order for the consistency

File: kornia/feature/efficient_loftr/backbone/repvgg.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+        out = self.stage0(x)
+        for stage in (self.stage1, self.stage2, self.stage3):
+            for block in stage:
+                if self.use_checkpoint:
+                    out = checkpoint.checkpoint(block, out)
+                else:
+                    out = block(out)
+        out = self.gap(out)
+        out = out.view(out.size(0), -1)
+        out = self.linear(out)
```
Comment (ducha-aiki) [inline kornia/feature/efficient_loftr/backbone/repvgg.py:308]: Please, delete and clean up - the linear or gap were never definer in init

File: kornia/feature/efficient_loftr/utils/coarse_matching.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        # normalize
+        feat_c0, feat_c1 = (feat / feat.shape[-1] ** 0.5 for feat in [feat_c0, feat_c1])
+
+        if self.fp16matmul:
+            sim_matrix = torch.einsum("nlc,nsc->nls", feat_c0, feat_c1) / self.temperature
+            del feat_c0, feat_c1
+            if mask_c0 is not None and mask_c1 is not None:
+                sim_matrix = sim_matrix.masked_fill(~(mask_c0[..., None] * mask_c1[:, None]).bool(), -1e4)
+        else:
+            with torch.autocast(enabled=False, device_type="cuda"):
```
Comment (ducha-aiki) [inline kornia/feature/efficient_loftr/utils/coarse_matching.py:79]: please, unhardcode cuda

File: kornia/feature/efficient_loftr/eloftr_module/linear_attention.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+from typing import List, Optional
+
+import torch
+import torch.nn.functional as F
+from torch import Tensor, nn
+
+if hasattr(F, "scaled_dot_product_attention"):
+    FLASH_AVAILABLE = True
+    from torch.backends.cuda import sdp_kernel
```
Comment (ducha-aiki) [inline kornia/feature/efficient_loftr/eloftr_module/linear_attention.py:31]: deprecated since
  torch 2.3; `torch.nn.attention.sdpa_kernel` is the replacement

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
There are still major blockers, please fix them.
